doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +962,563 @@ tactic
-      right (with respect to the raw printing obtained by setting the
-      :flag:`Printing All` flag).
+      .. coqtop:: all
 
-   .. example::
+         match True with
+         | True => idtac "branch 1"; fail
+         | _ => idtac "branch 2"
+         end.
 
-      .. coqtop:: all abort
+.. _match_vs_multimatch_ex:
 
-         Ltac f x :=
-           match x with
-             context f [S ?X] =>
-             idtac X;                    (* To display the evaluation order *)
-             assert (p := eq_refl 1 : X=1);    (* To filter the case X=1 *)
-             let x:= context f[O] in assert (x=O) (* To observe the context *)
+   .. example:: Comparison of match and multimatch
+
+      :tacn:`match` tactics are only evaluated once, whereas :tacn:`multimatch`
+      tactics may be evaluated more than once if the following constructs trigger backtracking:
+
+      .. coqtop:: all fail
+
+         Fail match True with
+         | True => idtac "branch 1"
+         | _ => idtac "branch 2"
+         end ;
+         match True with
+         | True => idtac "branch A"; fail
+         end.
+
+      .. coqtop:: all fail
+
+         Fail multimatch True with
+         | True => idtac "branch 1"
+         | _ => idtac "branch 2"
+         end ;
+         match True with
+         | True => idtac "branch A"; fail
+         end.
+
+.. _match_with_holes_ex:
+
+   .. example:: Matching a pattern with holes
+
+      Notice the :tacn:`idtac` prints ``(z + 1)`` while the :tacn:`pose` substitutes
+      ``(x + 1)``.
+
+      .. coqtop:: in reset
+
+         Goal True.
+
+      .. coqtop:: all
+
+           match constr:(fun x => (x + 1) * 3) with
+           | fun z => ?y * 3 => idtac "y =" y; pose (fun z: nat => y * 5)
            end.
+
+.. _match_term_context_ex:
+
+   .. example:: Multiple matches for a "context" pattern.
+
+      Internally "x <> y" is represented as "(not x y)", which produces the
+      first match.
+
+      .. coqtop:: in reset
+
+         Ltac f t := match t with
+                    | context [ (not ?t) ] => idtac "?t = " t; fail
+                    | _ => idtac
+                    end.
          Goal True.
-         f (3+4).
+
+      .. coqtop:: all
+
+         f ((not True) <> (not False)).
 
 .. _ltac-match-goal:
 
-Pattern matching on goals
-~~~~~~~~~~~~~~~~~~~~~~~~~
+Pattern matching on goals and hypotheses: match goal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-We can perform pattern matching on goals using the following expression:
+.. tacn:: @match_key {? reverse } goal with {? %| } {+| @goal_pattern => @ltac_expr } end
+   :name: lazymatch goal; match goal; multimatch goal
 
-.. we should provide the full grammar here
+   .. insertprodn goal_pattern match_hyp
 
-.. tacn:: match goal with {+| {+, @context_hyp} |- @cpattern => @ltac_expr } | _ => @ltac_expr end
-   :name: match goal
+   .. prodn::
+      goal_pattern ::= {*, @match_hyp } |- @match_pattern
+      | [ {*, @match_hyp } |- @match_pattern ]
```

Comment:
I don't think the meaning of the square brackets is explained anywhere...

Reply:
Note that the square brackets have no meaning, and are purely aesthetic (much like the initial `|` in inductive constructor cases and in `match` statements).

Reply:
OK thanks, this should probably be written down somewhere.

Reply:
What do the square brackets mean in `@name := {? [ @match_pattern ] : } @match_pattern`?  That doesn't look like it's just syntactic sugar.

Reply:
Oh, I see in the next comment.